### PR TITLE
Add advanced_options_config for compute_security_policy

### PR DESCRIPTION
**Pull request by wyardley**
Add advanced_options_config for compute_security_policy

Fixes hashicorp/terraform-provider-google#9939
Fixes hashicorp/terraform-provider-google#10497

If this PR is for Terraform, I acknowledge that I have:

- [ ] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [ ] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-downstream-tools), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [ ] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/third_party/terraform/tests) (for handwritten resources or update tests).
- [ ] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/main/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [ ] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
compute: Added `advanced_options_config` to `google_compute_security_policy`
```

### CODE REVIEW

**Comment by wyardley** [comment]
Hi @megan07 -- haven't 100% verified local build / acceptance tests (and may be a `go fmt` issue possibly still), but wanted to get this up (let me know if I should convert to draft). Appreciate any help / suggestions.

File: mmv1/third_party/terraform/resources/resource_compute_security_policy.go.erb [hunk 1]
```diff
@@ -289,7 +289,7 @@ func resourceComputeSecurityPolicy() *schema.Resource {
 				Description: `The URI of the created resource.`,
 			},
 
-		<% unless version == 'ga' -%>
+		<%- unless version == 'ga' -%>
```
Comment (wyardley) [inline mmv1/third_party/terraform/resources/resource_compute_security_policy.go.erb:292]: This creates extra spaces on the next line if `-` isn't set to suppress the leading tabs. Best I can tell, it still generates the GA code just fine.

**Comment by megan07** [comment]
```
Error: expected advanced_options_config.0.json_parsing to be one of [STANDARD DISABLED], got ENABLED
on terraform_plugin_test.tf line 7, in resource "google_compute_security_policy" "policy":
    json_parsing = "ENABLED"
Error: expected advanced_options_config.0.log_level to be one of [NORMAL VERBOSE], got DEBUG
on terraform_plugin_test.tf line 8, in resource "google_compute_security_policy" "policy":
    log_level    = "DEBUG"
--- FAIL: TestAccComputeSecurityPolicy_withAdvancedOptionsConfig (2.83s)
```

**Comment by wyardley** [comment]
> Error: expected advanced_options_config.0.json_parsing to be one of [STANDARD DISABLED], got ENABLED
on terraform_plugin_test.tf line 7, in resource "google_compute_security_policy" "policy":
>    json_parsing = "ENABLED"

Good call - updating this now. I had thought that (confusingly) "disabled" was actually the non-default option, but reading through the docs again, I think `STANDARD` is, confusingly, _not_ standard, since it's disabled by default? 🤔 
https://cloud.google.com/armor/docs/configure-security-policies

**Comment by megan07** [comment]
Yes, I missed this as well. It looks like `log_level` also had an error as `DEBUG` doesn't seem to be an accepted value either.

**Comment by wyardley** [comment]
> Yes, I missed this as well. It looks like log_level also had an error as DEBUG doesn't seem to be an accepted value either.

🤦 ok, yes, you're right. Fixed that too (right in the docs, wrong in the integration test).

I'm still trying to double-check the exact behavior of "STANDARD" vs "DISABLED". As confusing as it is, I _think_ DISABLED is standard and STANDARD is non-standard.

File: mmv1/third_party/terraform/resources/resource_compute_security_policy.go.erb [hunk 2]
```diff
@@ -322,6 +322,30 @@ func resourceComputeSecurityPolicy() *schema.Resource {
...
+				Type: schema.TypeList,
+				Optional: true,
+				Description: `Advanced Options Config of this security policy.`,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"json_parsing": {
+								Type: schema.TypeString,
+								Optional: true,
+								Default: "STANDARD",
```
Comment (megan07) [inline mmv1/third_party/terraform/resources/resource_compute_security_policy.go.erb:335]: I'm now a little confused as well :) but if `DISABLED` is the default, we probably want that to match here.

Comment (wyardley) [reply mmv1/third_party/terraform/resources/resource_compute_security_policy.go.erb]: You're right - got it. Updated!

Also checking with someone on the GCP Slack to verify that I'm understanding correctly which is which, but if you've got any contacts internally at Hashi or Google who can help, that would also be great.

**Comment by wyardley** [comment]
Also confirmed with someone who should be well informed on GCP Slack that:

> standard enables the json parsing

and

> disabled (default)  turns json parsing off

**Comment by wyardley** [comment]
@megan07 in the normal case where the user has the default settings (i.e., they don't have verbose logging, and have JSON parsing disabled), will this trigger "resource changed outside of terraform" in the state the first time they plan with the updated provider, or will this be completely a noop for people who have default settings?

**Comment by megan07** [comment]
It should be a complete noop, but I can test that locally to be sure.

The tests results are below.
```
provider_test.go:309: ImportStateVerify attributes not equivalent. Difference is shown below. Top is actual, bottom is expected.
(map[string]string) {
}
(map[string]string) (len=4) {
(string) (len=25) "advanced_options_config.#": (string) (len=1) "1",
(string) (len=27) "advanced_options_config.0.%": (string) (len=1) "2",
(string) (len=38) "advanced_options_config.0.json_parsing": (string) (len=8) "STANDARD",
(string) (len=35) "advanced_options_config.0.log_level": (string) (len=7) "VERBOSE"
}
--- FAIL: TestAccComputeSecurityPolicy_withAdvancedOptionsConfig (32.05s)
```

**Comment by wyardley** [comment]
> The tests results are below.

@megan07 thanks. So it looks like it's just getting nothing back with those parameters set?

Can you verify that the overall data structures look correct to you around line 326? I reviewed it against other attributes, and I _think_ it's right, but my golang is a little rusty.

I took another look at 
https://cloud.google.com/compute/docs/reference/rest/v1/securityPolicies
and it looks like (other than being one level more deeply nested), `adaptiveProtectionConfig` should be a good rough example. I assume the snake => camel case equivalency is builtin.

Will also fix the bad indent on line 333, though `go fmt` ends up fixing it in the rendered template anyway.

BTW, still can't build locally (which might help with testing if there's an easy way to also include the built beta provider), but for reasons that I don't _think_ have to do with this?

1. `scripts/diff.go` isn't formatted, and doesn't _get_ formatted by `go fmt`, though I can format it manually
2. a lot of these type of errors when trying to build the beta provider (tried doing `go mod download`, but no difference). Should I see ~ 728 files changed from `main` after building beta provider?
```
google-beta/access_context_manager_operation.go:48:169: undefined: time
google-beta/access_context_manager_operation.go:59:123: undefined: time
google-beta/active_directory_operation.go:50:173: undefined: time
[...]
```

**Comment by wyardley** [comment]
Also, I had been thinking this was a beta feature, but since this is in the and can be triggered via the (non-beta) `gcloud` CLI, I'm thinking maybe this shouldn't be gated by the ga guard clause after all. Let me make that change.

**Comment by megan07** [comment]
@wyardley - first, i'm so sorry, I must have really been sleeping when I reviewed this, I missed a huge piece 🤦‍♀️ 
You'll need to update the requests and set it in state. See the examples below.
[Create](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/third_party/terraform/resources/resource_compute_security_policy.go.erb#L375), [Read](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/third_party/terraform/resources/resource_compute_security_policy.go.erb#L455), [Update](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/third_party/terraform/resources/resource_compute_security_policy.go.erb#L487) (if its updateable)

I actually missed that a lot when I first started out too, typically I'll take a similar field (as you mentioned `adaptive_protection_config`) and do a search on the file and basically copy and paste all pieces and work from there.

As for compiling the tests, do you have a [downstream repo](https://github.com/hashicorp/terraform-provider-google-beta) cloned that you're building too? I guess I'd have to see what you're running and how it's set up to be able to help.

**Comment by wyardley** [comment]
Thanks! 🙏 Yea, first time contributing here and a lot of moving parts. Will work on getting those updated.

Testing the tf Google provider regular / beta downstream repos with the examples in the readme, so yes, I have those cloned.

**Comment by wyardley** [comment]
@megan07 I'm sure I may have messed something up or have some copy / pasta, but can you see if this is missing anything high level, or has any obvious mistakes?

**Comment by wyardley** [comment]
Also, had missed installing `goimports` from `tools/doctor` - will see if that helps with my local build issues. Looks like this is failing lint in CI

**Comment by wyardley** [comment]
Basically, what I'm seeing is that running, e.g., 
```
% make terraform VERSION=ga OUTPUT_PATH="$GOPATH/src/github.com/hashicorp/terraform-provider-google"                 
% make terraform VERSION=beta OUTPUT_PATH="$GOPATH/src/github.com/hashicorp/terraform-provider-google-beta"                         
```           

against current trunk generates a bunch of changes / warnings that aren't related to my changes. I can try building just this product to see if that helps.

**Comment by wyardley** [comment]
@megan07 so I can test it locally, and I can functionally test with the built provider and actual resources 🎉
Thanks for the help with that! Some progress

However, I am getting a permadiff with defaults, and maybe an issue toggling back to default settings as well.
```
  # module.waf_rules.google_compute_security_policy.policy will be updated in-place
  ~ resource "google_compute_security_policy" "policy" {
        id          = "projects/xyz/global/securityPolicies/waf-rules"
        name        = "waf-rules"
        # (4 unchanged attributes hidden)

      - advanced_options_config {
          - json_parsing = "DISABLED" -> null
          - log_level    = "NORMAL" -> null
        }

        # (8 unchanged blocks hidden)
    }
```

So perhaps some issues with the flattener and / or overall code, but at least I'm in a better position to test with fewer round trips.

**Comment by megan07** [comment]
@wyardley great! so i took a look at this more now. yes, I think we're missing a piece in update. similar to the [`description` update](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/third_party/terraform/resources/resource_compute_security_policy.go.erb#L487-L490), I think you'll want something like this:

```
	if d.HasChange("advanced_options_config") {
		securityPolicy.AdvancedOptionsConfig = expandSecurityPolicyAdvancedOptionsConfig(d.Get("advanced_options_config").([]interface{}))
		securityPolicy.ForceSendFields = append(securityPolicy.ForceSendFields, "AdvancedOptionsConfig")
	}
```

as for the permadiff, that's going to be an API-side issue, it appears that the API doesn't return the `AdvancedOptionsConfig` unless it is set, which is more inline to there _not_ being default values in the block, other than the documentation that says there is. does that make sense?

you can test your update function by adding extra steps to your test, if you'd like. let me know if you need more direction to do that. thanks again for contributing! we really appreciate it!

**Comment by wyardley** [comment]
> I think you'll want something like this:

Thanks. I was messing with `expandSecurityPolicyAdvancedOptionsConfig()` itself, but let me see if updating the update code to include the forcesendfields bit works. It is confusing, since once _one_ option is set, we want to make sure that the _other_ option is the default if unset, but if neither is set, we want to just try to leave it unset / clear out that field.

I think this basically works for most normal uses cases, however, once it's been set once, there will still be a permadiff from trying to do this:
```diff
      - advanced_options_config {
          - json_parsing = "STANDARD" -> null
          - log_level    = "NORMAL" -> null
        }
```
It does seem like the API leaves it present once set, even if returned to default values, so this may be unavoidable (though someone could resolve it by destroying / recreating it). Is there a way to make tf not try to manage it when set to default values, or is this not worth it?

```
rita~% gcloud compute security-policies update waf-rules \
    --log-level=normal --json-parsing=disabled --project xxx
Updated [https://www.googleapis.com/compute/v1/projects/xxx/global/securityPolicies/waf-rules].
% gcloud compute security-policies describe waf-rules --project xxx | yq .advancedOptionsConfig                                                  
jsonParsing: DISABLED
logLevel: NORMAL
```
> you can test your update function by adding extra steps to your test, if you'd like. let me know if you need more direction to do that.

I was thinking the same thing - it would probably be good to have a test for the update as well - I found a few examples of similar things, but let me know if there's an obvious good example to base it on. Would it be a separate test, or a multi-stage / templated out test?

Basically, start with one non-default value, then update the other parameter as well or something, that should be a good test, right?

Appreciate the help and direct example!

**Comment by megan07** [comment]
> It is confusing, since once one option is set, we want to make sure that the other option is the default if unset, but if neither is set, we want to just try to leave it unset / clear out that field.

I'm going to run this by someone else who understands the SDKv2's handing of this better than I do, but I think what you have is the best we can do right now. I'll let you know if I hear otherwise.

As for the update test, I would keep it one test. I looked into it a bit, I think I would do this
```
			{
				Config: testAccComputeSecurityPolicy_basic(spName),
			},
			{
				ResourceName:      "google_compute_security_policy.policy",
				ImportState:       true,
				ImportStateVerify: true,
			},
			{
				Config: testAccComputeSecurityPolicy_withAdvancedOptionsConfig(spName),
			},
			{
				ResourceName:      "google_compute_security_policy.policy",
				ImportState:       true,
				ImportStateVerify: true,
			},
```

And then just change the `type` in the basic test to be `CLOUD_ARMOR`. We usually write our tests this way, but for some reason the other tests here aren't using the basic test, so it should be safe to change the `type` on that without breaking anything else. Let me know if you have other questions!

**Comment by wyardley** [comment]
@megan07 Ok - if that's good enough (updating to include these fields), I think that's fairly simple / straightforward. Added.

Let me know if this looks good.

**Comment by wyardley** [comment]
Slightly misunderstood and originally added this to the basic config -- fixing and repushing with that in `TestAccComputeSecurityPolicy_withAdvancedOptionsConfig()` instead.

Also, I don't see `CLOUD_ARMOR_EDGE` type tested anywhere else in the file, so changing the basic config might remove a test case of that; let me know if that's an issue and I can make a different starting config to use, but maybe that should be its own separate test case?

**Comment by megan07** [comment]
@wyardley looks good. we could probably add `CLOUD_ARMOR_EDGE` to one of the other tests if you want (I ran it locally on the `testAccComputeSecurityPolicy_withRule`, `testAccComputeSecurityPolicy_updateSamePriority`, and `testAccComputeSecurityPolicy_update` configurations, all part of the `TestAccComputeSecurityPolicy_update` test, and it passed fine.)

You had brought up "resource changed out of terraform" for current configs, and as I was looking into that I came across a bug - I think in the API. You'll see it if you add another step to your test to go back to the basic config. Basically, removing this block won't reset the values to the default, it will keep the current values. I tried updating [this line](https://github.com/GoogleCloudPlatform/magic-modules/pull/5851/files#diff-7f40a6d1420bc5b6fdfefde5279bb9ff7d5088b470f4133930e89b26c4110377R784) to return an empty `SecurityPolicyAdvancedOptionsConfig` object, but the API seems to ignore that. I will file a bug upstream to see if they can address it.

**Comment by wyardley** [comment]
> You had brought up "resource changed out of terraform" for current configs, and as I was looking into that I came across a bug - I think in the API. You'll see it if you add another step to your test to go back to the basic config. Basically, removing this block won't reset the values to the default, it will keep the current values.

I _think_ I saw that in some of my testing. I'm not sure if it's a bug in the API itself, though, because I think it's possible to toggle the versions (just not to remove the block entirely) with the CLI and have them "stick". Could the problem be because we're not explicitly setting them back to defaults if / when we remove the configuration? Since the API doesn't seem to let you remove `advancedOptionsConfig` at all once it's been set (without destroying the rule), maybe we'd have to either force recreation (this would not be ideal, because I think you can't remove a policy that's in use) on the resource when the block is removed, or update the code to set default values if
a) advancedOptionsConfig exists
b) no values are provided?

**Comment by wyardley** [comment]
> we could probably add CLOUD_ARMOR_EDGE to one of the other tests if you want

I'm inclined to do as little work as required here 😉

**Comment by megan07** [comment]
@wyardley We could just reset it to the defaults ourselves, but that's not ideal and not necessarily how we'd expect the API to work. I've submitted a bug with google (https://issuetracker.google.com/issues/227492054) and raised it internally. We'll see what they say and can decide if we need to work around it or if it's a bug they'll fix themselves

**Comment by wyardley** [comment]
> We could just reset it to the defaults ourselves, but that's not ideal and not necessarily how we'd expect the API to work.

Agree. thanks for making the issue tracker ticket!

**Comment by wyardley** [comment]
Rebased again, just to keep it up to date

q: normally I've seen like this take quite a while to get a response, let alone to get fixed. Is it worth either forcing the defaults or shipping it with the limitation that once you've set it, you need to set it explicitly?

**Comment by wyardley** [comment]
@megan07 did you hear anything back from the people you spoke to?

**Comment by megan07** [comment]
@wyardley thanks for rebasing! I'm sorry for my delayed response on this, I've got another priority that I've been trying to wrap up and after that I will follow up with the team and see what they say. I figured the time it takes me to work on my other issue will hopefully give them ample time to look into the request. I will come back to it though (hopefully shortly). Thanks for your patience!

**Comment by wyardley** [comment]
@megan07 heard anything more back? Doesn't seem to have been accepted / assigned.

File: mmv1/third_party/terraform/resources/resource_compute_security_policy.go.erb [hunk 3]
```diff
@@ -512,6 +545,11 @@ func resourceComputeSecurityPolicyUpdate(d *schema.ResourceData, meta interface{
 		securityPolicy.ForceSendFields = append(securityPolicy.ForceSendFields, "Description")
 	}
 
+	if d.HasChange("advanced_options_config") {
+		securityPolicy.AdvancedOptionsConfig = expandSecurityPolicyAdvancedOptionsConfig(d.Get("advanced_options_config").([]interface{}))
+		securityPolicy.ForceSendFields = append(securityPolicy.ForceSendFields, "AdvancedOptionsConfig")
```
Comment (megan07) [inline mmv1/third_party/terraform/resources/resource_compute_security_policy.go.erb:550]: Let's try this?
```suggestion
		securityPolicy.ForceSendFields = append(securityPolicy.ForceSendFields, "AdvancedOptionsConfig", "advancedOptionsConfig.jsonParsing", "advancedOptionsConfig.logLevel")
```

File: mmv1/third_party/terraform/tests/resource_compute_security_policy_test.go.erb [hunk 4]
```diff
@@ -123,6 +123,36 @@ func TestAccComputeSecurityPolicy_update(t *testing.T) {
...
+				ImportStateVerify: true,
+			},
+			{
+				Config: testAccComputeSecurityPolicy_withAdvancedOptionsConfig(spName),
+			},
+			{
+				ResourceName:      "google_compute_security_policy.policy",
+				ImportState:       true,
+				ImportStateVerify: true,
+			},
```
Comment (megan07) [inline mmv1/third_party/terraform/tests/resource_compute_security_policy_test.go.erb:151]: Can we add this as the last step as well, just to catch the error?
```
			{
				Config: testAccComputeSecurityPolicy_basic(spName),
			},
			{
				ResourceName:      "google_compute_security_policy.policy",
				ImportState:       true,
				ImportStateVerify: true,
			},
```

**Review by megan07 — CHANGES_REQUESTED** [review]
Can we add that last step to the test, and just try the one change I suggested there (no promises it will work). If it doesn't work, and you can find another work around to get that last step to pass, we can merge this with a note of the workaround and a link to the issue we opened.

File: mmv1/third_party/terraform/tests/resource_compute_security_policy_test.go.erb (hunk 4 above)
Comment (wyardley) [reply mmv1/third_party/terraform/tests/resource_compute_security_policy_test.go.erb:151]: 👍 thanks for these.

**Comment by wyardley** [comment]
Not sure if the current test failure is expected?

Did some quick functional testing

1) If the value is, or has been, set (either with the provider or using the CLI), it seems to still generate a permadrift unless `advanced_options_config` is defined with its current / expected values.
2) if a new policy is created without  an `advanced_options_config` block, it seems to get defaults, as expected, and no weird diffs

So if I set it explicitly in the provider away from defaults with, e.g.,
```hcl
resource "google_compute_security_policy" "policy" {
  name        = "test"
  description = "test"

  advanced_options_config {
    json_parsing = "STANDARD"
    log_level    = "VERBOSE"
  }
}
```
and then remove the block entirely, Terraform claims to successfully modify the resource on apply, but leaves the value set to
```
% gcloud compute security-policies describe test --project myproject | yq .advancedOptionsConfig
jsonParsing: STANDARD
logLevel: VERBOSE
```
and then continues to show a permadiff
```
  # google_compute_security_policy.policy will be updated in-place
  ~ resource "google_compute_security_policy" "policy" {
        id          = "projects/myproject/global/securityPolicies/test"
        name        = "test"
        # (5 unchanged attributes hidden)

      - advanced_options_config {
          - json_parsing = "STANDARD" -> null
          - log_level    = "VERBOSE" -> null
        }

        # (1 unchanged block hidden)
    }
```

**Comment by wyardley** [comment]
Hi @megan07  - can you take another look when you get a chance?

**Comment by megan07** [comment]
Hi @wyardley , yea, that is not what we would expect. I guess we could potentially force it to do what we'd expect, right? If the configuration is removed, we can hard code it to use the defaults? What are your thoughts? (Sorry, I have context switched a lot since looking at this) I ran into a similar resource the other day and it seems as thought that's what that resource did. It was also bugged (which is why I was digging into it), but that was the attempt at least. If we can get that test to pass without a permadiff, I can merge it :) Thanks! And sorry again for the delay on this!

**Comment by wyardley** [comment]
> yea, that is not what we would expect. I guess we could potentially force it to do what we'd expect, right? If the configuration is removed, we can hard code it to use the defaults?

Only problem is that then I think we'd _always_ be hard-coding it? 
Probably it is an edge case that someone would set the config and then remove it completely, but it is definitely annoying.

> If we can get that test to pass without a permadiff, I can merge it 

I don't think I can see the (CI) test results, right? Can you confirm that you're seeing a permadiff there? I was mostly playing around with a local functional test.

Happy to tweak some more things if you've got some suggestions for what to try next.

File: mmv1/third_party/terraform/resources/resource_compute_security_policy.go.erb [hunk 5]
```diff
@@ -768,6 +806,31 @@ func flattenMatchExpr(match *compute.SecurityPolicyRuleMatcher) []map[string]int
 	return []map[string]interface{}{data}
 }
 
+func expandSecurityPolicyAdvancedOptionsConfig(configured []interface{}) *compute.SecurityPolicyAdvancedOptionsConfig {
+	if len(configured) == 0 || configured[0] == nil {
+		return nil
```
[the remark this answers is not in the document]
Comment (megan07) [reply mmv1/third_party/terraform/resources/resource_compute_security_policy.go.erb:811]: Ah, yes, ok, so the drawback to this is if `unset` differs from `default`, which it sounds like it does in this case. This will assume that unset will ALWAYS default.

There shouldn't be a diff, I don't think? Because the state will match the default. You might have to make the parent (`advanced_options_config`) `Optional` and `Computed` maybe, but I'd have to do some playing around with that to confirm it. However, an empty configuration, with defaults, should not conflict with defaults in state. It's possible I'm missing something here, as usually I like to test something out - I just don't have that time at the current moment, but I can mess with it more later perhaps.

The other option would be to make them `Optional` and `Computed`, but then we're going to still hit the upstream bug, I'd imagine - it would probably just get rid of the diff (which, I suppose is fine - we'd be mimicking the API at least).

Would you mind trying those options? If they both work, pick which solution you like best knowing the drawbacks.

[the remark this answers is not in the document]
Comment (megan07) [reply mmv1/third_party/terraform/resources/resource_compute_security_policy.go.erb:811]: Ok, update! They both pass, however, trying to hardcode the default doesn't work as intended (i think it's a long known SDK issue that i still haven't figured out how to work around), so I think our best solution would be to make everything (`advanced_options_config` and its children) `Optional` and `Computed`. This will mimic the API so that the bug will still be present, and removing the configuration won't do anything, but our permadiff will go away. The only way to reset it to the defaults would be to update the configuration to the appropriate value. Does this work for you?

[the remark this answers is not in the document]
Comment (wyardley) [reply mmv1/third_party/terraform/resources/resource_compute_security_policy.go.erb:811]: > Ok, update! They both pass,

Thanks for testing it - really appreciate it.

> The only way to reset it to the defaults would be to update the configuration to the appropriate value. Does this work for you?

If I'm understanding you right, yes, I think that works. To summarize / paraphrase, I think the behavior is
* defaults / advanced options _not_ set, creating new resource or updating provider for existing resource: no changes
* options set explicitly away from (or to) default: options will be set as requested, however, once removed, the configuration would _then_ need to have (both) the default values (or whatever values they want) explicitly set in the config.

[the remark this answers is not in the document]
Comment (megan07) [reply mmv1/third_party/terraform/resources/resource_compute_security_policy.go.erb:811]: Correct!

[the remark this answers is not in the document]
Comment (wyardley) [reply mmv1/third_party/terraform/resources/resource_compute_security_policy.go.erb:814]: thanks. Yes, I think that's the best we can do.

@megan07 I rebased, added `Computed: true,` to both the parent and children, and pushed up an update. I have not yet tested this locally. Do you want to confirm that this is what you tested and what you were suggesting?

[the remark this answers is not in the document]
Comment (wyardley) [reply mmv1/third_party/terraform/resources/resource_compute_security_policy.go.erb:812]: I also had to remove the defaults - otherwise, failed with this error:
```
│ 1 error occurred:
│ 	* resource google_compute_security_policy: json_parsing: Default must be nil if computed
```

**Comment by wyardley** [comment]
Ok, this seems to _pretty much_ work @megan07:

If I start with this:
```hcl
resource "google_compute_security_policy" "test" {
  name = "test"
}
```

I get the following plan
```
  # google_compute_security_policy.test will be created
  + resource "google_compute_security_policy" "test" {
      + fingerprint = (known after apply)
      + id          = (known after apply)
      + name        = "test"
      + project     = (known after apply)
      + self_link   = (known after apply)
      + type        = (known after apply)

      + advanced_options_config {
          + json_parsing = (known after apply)
          + log_level    = (known after apply)
        }
[...]
```
and it applies cleanly. Also, it still appears to be unset in the backend:
```
% gcloud compute security-policies describe test --project xyz | yq .advancedOptionsConfig                 
null
```

If I add this to the config
```hcl
  advanced_options_config {
    json_parsing = "STANDARD"
    log_level    = "VERBOSE"
  }
```

the plan looks like:
```
  # google_compute_security_policy.test will be updated in-place
  ~ resource "google_compute_security_policy" "test" {
        id          = "projects/xyz/global/securityPolicies/test"
        name        = "test"
        # (4 unchanged attributes hidden)

      + advanced_options_config {
          + json_parsing = "STANDARD"
          + log_level    = "VERBOSE"
        }

        # (1 unchanged block hidden)
    }
```

and I see this:
```
% gcloud compute security-policies describe test --project xyz | yq .advancedOptionsConfig
jsonParsing: STANDARD
logLevel: VERBOSE
```

One caveat (which I _think_ we said was expected, but just want to confirm) is that if I remove `advanced_options_config` entirely from the config, it seems to silently leave it alone vs. forcing it back to defaults. I'm not sure if we need to call that out in the commit message, release notes, or elsewhere, but I think this is probably the least annoying config that will more or less work for most use cases.

File: mmv1/third_party/terraform/resources/resource_compute_security_policy.go.erb (hunk 5 above)
[the remark this answers is not in the document]
Comment (wyardley) [reply mmv1/third_party/terraform/resources/resource_compute_security_policy.go.erb:812]: Commented below in the main comments with how this behaves with some simple functional tests.

**Comment by wyardley** [comment]
Let me know if that crash from the VCR tests is related or not to these changes.

Also, question: now that defaults aren't set, do we need to change the docs about defaults at all? Still probably want to make it clear which options are the default behavior, but I'm not sure if the wording should be updated / changed.

**Comment by megan07** [comment]
The test failures don't appear to be related so far as I can tell - I'm not sure what happened with tests as I'm seeing some tests failing on other PRs as well, I'll dig into that today.

As for docs, I think I'd recommend removing any indications of `Default` as Terraform isn't defaulting to anything, and the underlying API can switch its defaults at any point, which would make our docs out of date. Especially with the fact that the defaults don't work as we're expecting anyway. How do you feel about that?

**Comment by megan07** [comment]
@wyardley ok, sorry for the back and forth on this. I talked to the team and we're in agreement we can throw them back in. I think you could just append a sentence to each with `Defaults to "VAL"`. Thanks!

**Comment by wyardley** [comment]
> I think you could just append a sentence to each with Defaults to "VAL". Thanks!

Ok, how's this look now?
